src/mlpack/methods/lsh/lsh_search_impl.hpp

```diff
@@ -339,12 +340,301 @@ void LSHSearch<SortPolicy>::BaseCase(const size_t queryIndex,
         referenceIndex, distance);
 }
 
+
+// Compare class for <double, size_t> pair, used in GetAdditionalProbingBins.
+class CompareGreater
+{
+  public:
+    bool operator()(
+        std::pair<double, size_t> p1,
+        std::pair<double, size_t> p2){
+      //only compare the double values
+      return p1.first > p2.first;
+    }
+};
+
+//Returns the score of a perturbation vector generated by perturbation set A.
+//The score of a pertubation set (vector) is the sum of scores of the
+//participating actions.
+inline double perturbationScore(
+              const std::vector<size_t> &A,
+              const arma::vec &scores)
+{
+  double score = 0.0;
+  for (size_t i = 0; i < A.size(); ++i)
+    score+=scores[A[i]];
+  return score;
+}
+
+// Inline function used by GetAdditionalProbingBins. The vector shift operation
+// replaces the largest element of a vector A with (largest element) + 1.
+inline void perturbationShift(std::vector<size_t> &A)
+{
+  size_t max_pos = 0;
+  size_t max = A[0];
+  for (size_t i = 1; i < A.size(); ++i)
+  {
+    if (A[i] > max)
+    {
+      max = A[i];
+      max_pos = i;
+    }
+  }
```

Comment:
Ah, sorry, I didn't see a later comment of yours.  If the `std::vector<bool>` solution isn't to your liking and it doesn't give any performance change, then it's up to you which implementation you want to go with.  Personally I'd probably pick the one that gives clearer code, but which one that is is up to you.